### ISSUE TITLE
Cant send ${KEY_ENTER}, ${KEY_TAB} etc

### ISSUE THREAD

**Issue by sglebs** [fhoeben/hsac-xebium-bridge#4]
Hi @fhoeben ,

Xebium interprets some special "keys" which allows us to send TAB, ENTER etc. This is quite important in many test cases. I am helping a person trying to use your xebium bridge but he can't get TAB, ENTER etc to work as with his xebium tests. Here's what I tried:

```
!| storyboard |
| start browser | firefox | on url | http://www.google.com |
| ensure | do | sendKeys | on | id=lst-ib| with | meleca !-${KEY_ENTER}-! |
```

Unfortunately it does not work. Is there a way to send ENTER with the xebium bridge or do I have to switch to the hsac calls to do that?

In my own fork of Xebium I added support for this syntax: %$KEY_ENTER

Basically I used Xebium's alias mechanism and pre-defined a few special ones. Here: https://github.com/xebia/Xebium/commit/0de55c5204315fd19370d2b8a83b1e19a621cba5

```
+    private void initSpecialKeysMapping() {
+        aliases.put("$KEY_BACKSPACE", Keys.chord(Keys.BACK_SPACE));
+        aliases.put("$KEY_TAB", Keys.chord(Keys.TAB));
+        aliases.put("$KEY_ENTER", Keys.chord(Keys.ENTER));
+        aliases.put("$KEY_RETURN", Keys.chord(Keys.RETURN));
+    }
```

Any suggestions?

**Comment by fhoeben** [comment]
I'm not aware that I do anything to how such a value is passed to Xebium. The only thing I can imagine that causes different behavior is that the bridge always checks whether an element is enabled and displayed before allowing you to interact with it.

Can you check with a debugger whether the value is passed to Xebium or what is causing the bridge to mess with it?

**Comment by fhoeben** [comment]
Does Xebium already have the key support you describe, or is that something that you added to your own fork? I use, as you know, Xebium 0.12 and I don't see it there.

You could create your own subclass to add it to the bridge, or use BrowserTest's `press` and `type` commands.

**Comment by sglebs** [comment]
I use my own fork. I guess I could fork your bridge to point at my fork... well, it won't work because I do not publish my fork in the mvn repositories. Hmmm...

**Comment by sglebs** [comment]
Xebium takes \10 for ENTER:

```
!| storyboard |
| start browser | firefox | on url | http://www.google.com |
| ensure | do | sendKeys | on | id=lst-ib| with | meleca \10 |
```

The problem is the other keys, which I had to add as well in my own fork

```
+        value = value.replace("\\08", Keys.BACK_SPACE); // sglebs fix
+        value = value.replace("\\09", Keys.TAB); // sglebs fix
```

**Comment by sglebs** [comment]
The alias approach will not work with your bridge even if I use my own fork. That's because Xebium's doOn is like this:

```
    public boolean doOn(final String command, final String target) {
        LOG.info("Performing | " + command + " | " + target + " |");
        return executeDoCommand(command, new String[] { unalias(target) });
    }
```

Note the call to unalias. On the other hand, your doOn does not honor any aliasing mechanism:

```
    @Override
    public boolean doOn(String command, String target) {
        boolean result;
        if ("waitForVisible".equals(command)) {
            result = ensureTargetVisible(target);
        } else if ("click".equals(command)
                    || "clickAndWait".equals(command)) {
            By by = targetToBy(target);
            WebElement element = getSeleniumHelper().findElement(by);
            return getBrowserTest().clickElement(element);
        } else {
            ensureReadyForDo(command, target);
            result = super.doOn(command, target);
        }
        return result;
    }
```

**Comment by sglebs** [comment]
Can I add the missing alias functionality to your HsacSeleniumDriverFixture and submit as a pull request? Can I just get the code from Xebium and my own and put it there?

**Comment by sglebs** [comment]
I put the pull request here: https://github.com/fhoeben/hsac-xebium-bridge/pull/5

This now works:

```
!| storyboard |
| start browser | firefox | on url | http://www.google.com |
| ensure | do | sendKeys | on | id=lst-ib| with | meleca |
| ensure | do | sendKeys | on | id=lst-ib| with | %$KEY_ENTER |
```

Xebium does not seem to support an alias inside a longer line. I would like to fix that in my fork and here too. I mean, this still does not work:

```
!| storyboard |
| start browser | firefox | on url | http://www.google.com |
| ensure | do | sendKeys | on | id=lst-ib| with | meleca %$KEY_ENTER |
```

I'd like to fix it too. But first let me know if you will take this pull request.

**Comment by fhoeben** [comment]
I incorporated your changes, and put in some additional updates that seemed required to me.

I would recommend you to put more effort to writing tests using BrowserTest (or your own subclass of it) than try to improve Xebium...

```
|storyboard|browser test         |
|open      |http://www.google.com|
|type      |meleca               |
|press enter                     |
```

**Comment by fhoeben** [comment]
In release 0.1.6

**Comment by sglebs** [comment]
yes, I just want a migration path for the existing tests first. it is getting pretty good!
